### PR TITLE
Feature/tipsets

### PR DESCRIPTION
**Pull request by masterjedy**
### Description of the Change
Tipsets feature implemented partially implemented.

### Possible Drawbacks 
TipsetKey function decodeKey is not implemented yet and I'm not sure that encodeKey is implemented correctly, because I don't know some details about encode/decode CIDs.
Tests are not implemented yet.
Need to remove copying of CID in Tipset::create after a PR in libp2p is merged.

### CODE REVIEW

File: core/primitives/tipset/tipset.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  bool operator==(const Tipset &lhs, const Tipset &rhs) {
+    if (lhs.blks_.size() != rhs.blks_.size()) return false;
+    for (size_t i = 0; i < lhs.blks_.size(); ++i) {
+      if (codec::cbor::getCidOfCbor(lhs.blks_[i])
+          != codec::cbor::getCidOfCbor(rhs.blks_[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
```
Comment (turuslan) [inline core/primitives/tipset/tipset.cpp]: Compare `BlockHeader` directly, define `operator==` if necessary.

File: core/primitives/tipset/tipset.cpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+namespace fc::common {
+  /**
+   * @brief comparator wrapper for optionals
+   * @tparam T underlying type
+   * @param lhs first instance
+   * @param rhs second instance
+   * @return true if they both don't have value or equal, false otherwise
+   */
+  template <class T>
+  bool operator==(const boost::optional<T> &lhs,
```
Comment (turuslan) [inline core/primitives/tipset/tipset.cpp]: Remove

File: core/primitives/tipset/tipset.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+  outcome::result<std::reference_wrapper<const block::BlockHeader>>
+  Tipset::getMinTicketBlock() const {
+    std::reference_wrapper<const block::BlockHeader> block = blks_[0];
+    if (!block.get().ticket.has_value()) {
+      return TipsetError::TICKET_HAS_NO_VALUE;
+    }
+    for (auto &b : blks_) {
+      if (!b.ticket.has_value()) {
+        continue;
```
Comment (turuslan) [inline core/primitives/tipset/tipset.cpp]:
```suggestion
        return TipsetError::TICKET_HAS_NO_VALUE;
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -49,7 +49,7 @@ jobs:
           sudo apt-get install -y ninja-build
         fi
 
-        sudo python3 -m pip install --upgrade pip
+        pip3 -V || sudo python3 -m pip install --upgrade pip
```
Comment (Harrm) [inline .github/workflows/ci.yml]: Why is that needed?

File: core/primitives/block/block.hpp [hunk 5]
```diff
@@ -31,8 +31,21 @@ namespace fc::primitives::block {
     Signature bls_aggregate;
     uint64_t timestamp;
     boost::optional<Signature> block_sig;
+    uint64_t fork_signaling;
   };
 
+  inline bool operator==(const BlockHeader &lhs, const BlockHeader &rhs) {
```
Comment (Harrm) [inline core/primitives/block/block.hpp]: *sigh* It will be great when the default operator  <=> is available.

File: core/primitives/cid/cid_of_cbor.hpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef CPP_FILECOIN_CORE_PRIMITIVES_CID_CID_OF_CBOR_HPP
+#define CPP_FILECOIN_CORE_PRIMITIVES_CID_CID_OF_CBOR_HPP
+
+#include "cid.hpp"
```
Comment (Harrm) [inline core/primitives/cid/cid_of_cbor.hpp]:
```suggestion
#include " primitives/cid/cid.hpp"
```

File: core/primitives/tipset/tipset.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  using fc::primitives::tipset::TipsetError;
+  switch (e) {
+    case (TipsetError::NO_BLOCKS):
+      return "Need to have at least one block to create tipset";
+    case TipsetError::MISMATCHING_HEIGHTS:
+      return "Cannot create tipset, mismatching blocks heights";
+    case TipsetError::MISMATCHING_PARENTS:
+      return "Cannot create tipset, mismatching block parents";
+    case TipsetError::TICKET_HAS_NO_VALUE:
+      return "optional ticket is not initialized";
```
Comment (Harrm) [inline core/primitives/tipset/tipset.cpp]:
```suggestion
      return "An optional ticket is not initialized";
```

File: core/primitives/tipset/tipset.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+  CID Tipset::getParentStateRoot() const {
+    return blks[0].parent_state_root;
+  }
+
+  BigInt Tipset::getParentWeight() const {
+    return blks[0].parent_weight;
+  }
+
+  bool Tipset::contains(const CID &cid) const {
+    for (auto &c : cids) {
```
Comment (Harrm) [inline core/primitives/tipset/tipset.cpp]: std::find?

File: core/primitives/tipset/tipset.cpp [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+  /**
+   * @brief compares two Tipset instances
+   * @param lhs first tipset
+   * @param rhs second tipset
+   * @return true if equal, false otherwise
+   */
+  bool operator==(const Tipset &lhs, const Tipset &rhs) {
+    if (lhs.blks.size() != rhs.blks.size()) return false;
+    for (size_t i = 0; i < lhs.blks.size(); ++i) {
```
Comment (Harrm) [inline core/primitives/tipset/tipset.cpp]: std::equal?

File: core/primitives/tipset/tipset.cpp [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+  outcome::result<boost::optional<ticket::Ticket>> Tipset::getMinTicket()
+      const {
+    OUTCOME_TRY(block, getMinTicketBlock());
+    return block.get().ticket;
+  }
+
+  uint64_t Tipset::getMinTimestamp() const {
+    auto timestamp = blks[0].timestamp;
+    for (auto &b : blks) {
```
Comment (Harrm) [inline core/primitives/tipset/tipset.cpp]: std::min_element?

File: core/primitives/tipset/tipset.hpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+ * @brief Outcome errors declaration
+ */
+OUTCOME_HPP_DECLARE_ERROR(fc::primitives::tipset, TipsetError);
+
+namespace fc::primitives::tipset {
+  /**
+   * @struct Tipset implemented according to
+   * https://github.com/filecoin-project/lotus/blob/6e94377469e49fa4e643f9204b6f46ef3cb3bf04/chain/types/tipset.go#L18
+   */
+  struct Tipset {
```
Comment (Harrm) [inline core/primitives/tipset/tipset.hpp]: It's weird that although it's a structure it has a lot of methods. Of course, they are getters so it's acceptable, but still low key unexpected.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (masterjedy) [reply .github/workflows/ci.yml]: When CI stopped working, it was the solution. Not to install newer version if current still works. Anyway, version 19 is ok, and we don't need to spend time installing 20-th version.

File: core/primitives/tipset/tipset.cpp (hunk 8 above)
Comment (masterjedy) [reply core/primitives/tipset/tipset.cpp]: ok
